pennylane/operation.py

```diff
@@ -1001,6 +1001,82 @@ def __matmul__(self, other):
 
         raise ValueError("Can only perform tensor products between observables.")
 
+    def _obs_data(self):
+        r"""Extracts the data from an Observable/Tensor."""
+        obs = Tensor(self).non_identity_obs
+        tensor = set()
+
+        for ob in obs:
+            parameters = tuple(param.tostring() for param in ob.parameters)
+            tensor.add((ob.name, ob.wires, parameters))
+
+        return tensor
+
+    def compare(self, other):
+        r"""Compares an `~.Observable` or a `~.Tensor` with another :class:`~.Hamiltonian`, Tensor, or Observable,
+        to determine if they are equivalent.
+
+        An Observable/Tensor and a Hamiltonian/Observable/Tensor are equivalent if they represent the same operator
+        (their matrix representations are equal), and they are defined on the same wires.
+
+        .. Warning::
+
+            The `compare()` method **cannot** check for equivalence between a `qml.Hermitian` Observable and an
+            equivalent Hamiltonian/Tensor/Observable written in terms of Pauli observables.
+            If this were the case, the matrix form of Hamiltonians/Tensors would need
+            to be calculated, which would drastically increase runtime.
+
+        Returns:
+            (bool): True if the Observable/Tensor and the other Hamiltonian/Observable/Tensor are equivalent,
+                    False otherwise.
+
+        **Examples**
+
+        >>> ob1 = qml.PauliX(0) @ qml.Identity(1)
+        >>> ob2 = qml.Hamiltonian([1], [qml.PauliX(0)])
+        >>> ob1.compare(ob2)
+        True
+
+        >>> ob1 = qml.PauliX(0)
+        >>> ob2 = qml.Hermitian(np.array([[0, 1], [1, 0]]), 0)
+        >>> ob1.compare(ob2)
+        False
+        """
+        if isinstance(other, (Tensor, Observable)):
+            return other._obs_data() == self._obs_data()
+        if isinstance(other, qml.Hamiltonian):
+            return other.compare(self)
+
+        raise ValueError(
+            "Can only compare an Observable/Tensor, and a Hamiltonian/Observable/Tensor."
+        )
+
+    def __add__(self, other):
+        r"""The addition operation between Observables/Tensors/qml.Hamiltonian objects.
+        """
+        if isinstance(other, (Observable, Tensor)):
+            return qml.Hamiltonian([1, 1], [self, other], simplify=True)
+
+        if isinstance(other, qml.Hamiltonian):
+            return other + self
+
+        return NotImplemented
```

Comment:
I would suggest instead changing this (and all other occurrences, both here and in the `Hamiltonian` class) to 
```suggestion
        raise ValueError(f"Cannot add Observable and {type(other)}")
```
The reason being is that it's very easy to programmatically include an addition/multiplication etc. and make a mistake with types (e.g., say you index into a list incorrectly). This way, an error is raised indicating to the user where the issue is, rather than failing silently.